src/typescript/frontend/src/pages/trade/[market_name].tsx

```diff
@@ -202,14 +207,100 @@ export default function Market({ allMarketData, marketData }: Props) {
+          const priceLevel: ApiPriceLevel = msg.data;
+          queryClient.setQueriesData(
+            ["orderbook", marketData.market_id],
+            (prevData: Orderbook | undefined) => {
+              if (prevData == null) {
+                return undefined;
+              }
+              if (priceLevel.side === "buy") {
+                for (const [i, lvl] of prevData.bids.entries()) {
+                  if (priceLevel.price === lvl.price) {
+                    return {
+                      bids: [
+                        ...prevData.bids.slice(0, i),
+                        { price: priceLevel.price, size: priceLevel.size },
+                        ...prevData.bids.slice(i + 1),
+                      ],
+                      asks: prevData.asks,
+                    };
+                  } else if (priceLevel.price > lvl.price) {
+                    return {
+                      bids: [
+                        ...prevData.bids.slice(0, i),
+                        { price: priceLevel.price, size: priceLevel.size },
+                        ...prevData.bids.slice(i),
+                      ],
+                      asks: prevData.asks,
+                    };
+                  }
+                }
+                return {
+                  bids: [
+                    ...prevData.bids,
+                    { price: priceLevel.price, size: priceLevel.size },
+                  ],
+                  asks: prevData.asks,
+                };
+              } else {
+                for (const [i, lvl] of prevData.asks.entries()) {
+                  if (priceLevel.price === lvl.price) {
+                    return {
+                      bids: prevData.bids,
+                      asks: [
+                        ...prevData.asks.slice(0, i),
+                        { price: priceLevel.price, size: priceLevel.size },
+                        ...prevData.asks.slice(i + 1),
+                      ],
+                    };
+                  } else if (priceLevel.price < lvl.price) {
+                    return {
+                      bids: prevData.bids,
+                      asks: [
+                        ...prevData.asks.slice(0, i),
+                        { price: priceLevel.price, size: priceLevel.size },
+                        ...prevData.asks.slice(i),
+                      ],
+                    };
+                  }
+                }
+                return {
+                  bids: prevData.bids,
+                  asks: [
+                    ...prevData.asks,
+                    { price: priceLevel.price, size: priceLevel.size },
+                  ],
+                };
+              }
+            }
+          );
         } else {
           // TODO
         }
       } else {
         // TODO
       }
     };
-  }, [marketData?.market_id, account?.address]);
+  }, [marketData, account?.address, queryClient]);
+
+  // TODO update to include precision when backend is updated
```

Comment:
Can we tag an issue here, or ideally start an issue so that we don't have a floating TODO?

Reply:
Done